### ISSUE TITLE
Probable bug in bidirectional iterators

### ISSUE THREAD

**Issue by KolmogorovSmirnov** [seqan/seqan#2300]
The implementation of goDown seems to be incorrect when the following conditions are met 

- The iterator uses an underlying bidirectional FM index 

- The thrid argument is Rev() (that is, we are going down the suffix tree, or equivalently expanding the pattern to the right)

- The second parameter is a string of many characters (as opposed to one character, or no second parameter

Consider this simple program , sorry for the formating

```c++
#include <vector>
#include <fstream>
#include <iostream>
#include <string>
#include <queue>
#include <algorithm>
#include <string.h>

#include <seqan/basic.h>
#include <seqan/sequence.h>
#include <seqan/stream.h>
#include <seqan/index.h>
using namespace std;
using namespace seqan;

typedef FastFMIndexConfig<void, uint32_t> TFastConfig;
typedef Index<String<Dna5>, BidirectionalIndex<FMIndex<void, TFastConfig> > > TIndex;
typedef Iter<TIndex, VSTree<TopDown<ParentLinks<> > > > MyiterT;

int main (int argc, const char* argv[]) {
    TIndex index;
    open(index, "XTEST");
   
    MyiterT it(index);
    
    goDown(it, Dna5String("AAAAT"), Rev());
    cerr << representative(it) << endl;
    goDown(it, Dna5String("CG"), Rev());
    cerr << representative(it) << endl;
    goUp(it);
    cerr << representative(it) << endl;
    goUp(it);
    cerr  << representative(it) << endl;
    goUp(it);
    cerr << representative(it) << endl;

    return 0;
}
```

The output in my case is 

AAAAT
AAAATCG
AAAAT
AAAAT

The fourth string should be AAA since it is the next node up in the suffix tree, and this is the problem.

This problem does not happen when goingDown by a single letter, or with the Fwd() tag as the third parameter. ~~Furtheremore, is seems that goUp can go several edges back, based on history as evidenced by the third line (two letters in this case). Is this also a bug, since the same does not happen with unidirectional iterators ?~~

**Comment by KolmogorovSmirnov** [comment]
Now I have a problem with simple iterators as well ! Can someone please tell me if this a genuine bug, or if I am simply not understanding the doc ? The program is as follows : 

#include <iostream>
#include <string.h>
#include <seqan/basic.h>
#include <seqan/sequence.h>
#include <seqan/stream.h>
#include <seqan/index.h>
using namespace std;
using namespace seqan;

typedef Index<String<Dna5> > TIndex;
typedef Iter<TIndex, VSTree<TopDown<ParentLinks<> > > > MyiterT;


int main (int argc, const char* argv[]) {



    
    String<Dna5> mytext("AAAATCG");
    TIndex index(mytext);
    MyiterT it(index);
    

    goDown(it, Dna5String("AAAAT"));
    cout << representative(it) << endl;
    goDown(it, Dna5String("CG"));
    cout << representative(it) << endl;
    goUp(it);
    cout << representative(it) << endl;
    goUp(it);
    cout  << representative(it) << endl;
    goUp(it);
    cout << representative(it) << endl;




    


    return 0;

}


The output is

AAAATCG
AAAATCG
AAA
AA
A

The output I expect is (assuming goUp goes up a single edge, but even if this was going several edges the previous output would be incorrect).
AAAAT
AAAATCG
AAAATC
AAAAT
AAAA

**Comment by KolmogorovSmirnov** [comment]
Please find below a cleaer code, that does not use an index saved on the hard drive : 

``
#include <iostream>
#include <string.h>
#include <seqan/basic.h>
#include <seqan/sequence.h>
#include <seqan/stream.h>
#include <seqan/index.h>
using namespace std;
using namespace seqan;

typedef FastFMIndexConfig<void, uint32_t> TFastConfig;
typedef Index<String<Dna5>, BidirectionalIndex<FMIndex<void, TFastConfig> > > TIndex;
typedef Iter<TIndex, VSTree<TopDown<ParentLinks<> > > > MyiterT;


int main (int argc, const char* argv[]) {



    
    String<Dna5> mytext("AAAATCG");
    TIndex index(mytext);
    MyiterT it(index);
    

    goDown(it, Dna5String("AAAT"), Rev());
    cout << representative(it) << endl;
    goDown(it,Dna5String("CG"), Rev());
    cout << representative(it) << endl;
    goUp(it);
    cout << representative(it) << endl;
    goUp(it);
    cout  << representative(it) << endl;
    goUp(it);
    cout << representative(it) << endl;



    return 0;

}

I have changed the line
 `typedef Index<String<Dna5>, BidirectionalIndex<FMIndex<void, TFastConfig> > > TIndex;`
because otherwise it throws a compilation error (rightfully since there would be a type mismatch).

The output is exactly the same as in the first post. How come AAAAT is shown twice, if there is a goUp between the two cerr << representative(it) ?

**Comment by cpockrandt** [comment]
Dear @KolmogorovSmirnov,

thanks for reporting it. I've looked into this and it seems to be a bug. I'll let you know as soon as I opened a pull request with a fix.

**Comment by cpockrandt** [comment]
Dear @KolmogorovSmirnov,

there's a pull request with a fix (see #2302). It might take a few days until this PR goes into the develop branch. You might want to checkout my commit instead until then.

**Comment by KolmogorovSmirnov** [comment]
Thanks for your answer. However I am quite confused now as what is the function goUp supposed to do. is it supposed to 

- goUp in the history (true undo operation)


- goUp in the tree (i.e get the parent node) ?

The behavior with bidirectional iterators, and your SEQAN_ASSERT_EQ,  indicates that is the first case.
But the behavior with unidirectional iterators indicates that is the second case, as evidenced by the output 

AAATCG
AAATCG
AAA
AA
A

Unless it was specifically designed to do different things with the two types of iterators ?
Of course, when going down by a single letter, both behaviors of goUp will do the same thing, even if the edge has many letters.

Thanks.

**Comment by cpockrandt** [comment]
I get the exact same behaviour with a unidirectional FM index. The behavior depens on the kind of index you use, i.e. tree or trie.

If you use an enhanced suffix array, goUp will go up an edge in suffix tree.
For FM indices, goUp might go up several edges if you called goDown with a string.

**Comment by KolmogorovSmirnov** [comment]
Thanks I think I understand now, but shouldn't the doc be updated since it says that for goUp

> Iterates up one edge to the parent in a tree.

which is non-true for FM indices ?

**Comment by smehringer** [comment]
Fixed by #2302
